src/lib/yang/data.lua

```diff
@@ -298,18 +306,18 @@ local function struct_parser(keyword, members, ctype)
          -- Scalar/array parser responsible for requiring whitespace
          -- after keyword.  Struct/table don't need it as they have
          -- braces.
-         local sub = assert(members[k], 'unrecognized parameter: '..k)
+         local sub = assert(expanded_members[k], 'unrecognized parameter: '..k)
          local id = normalize_id(k)
-         ret[id] = sub.parse(P, ret[id])
+         ret[id] = sub.parse(k, P, ret[id])
          P:skip_whitespace()
       end
-      for _,k in ipairs(keys) do
+      for k,_ in pairs(expanded_members) do
          local id = normalize_id(k)
-         ret[id] = members[k].finish(ret[id])
+         ret[id] = expanded_members[k].finish(ret[id], k)
       end
       return ret
    end
-   local function parse(P, out)
+   local function parse(k, P, out)
```

Comment:
I think instead of adding 'k' as the first parameter, it might be better to add it as the last parameter. This way, it's not necessary to make it `nil` explicitly in function calls (for instance, `parser.parse(nil, P, parser.init())`).

Another doubt is that `k` is not used in the function body. Is `k` meant to be `keyword`? If `k` is `keyword` and it can be nil, then I think a guard should be added to initialize k as empty in case it's nil (if not P:error('duplicate parameter: '..keyword) may cause the program to crash)) or add an assert to check keyword must exist in case of error.